<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Описание данных</a></span></li><li><span><a href="#Откроем-и-изучим-файл" data-toc-modified-id="Откроем-и-изучим-файл-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Откроем и изучим файл</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Разобьем-данные-на-выборки" data-toc-modified-id="Разобьем-данные-на-выборки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Разобьем данные на выборки</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Исследуйте-модели" data-toc-modified-id="Исследуйте-модели-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Исследуйте модели</a></span><ul class="toc-item"><li><span><a href="#Модель--Дерево-решений-(DecisionTreeClassifier)" data-toc-modified-id="Модель--Дерево-решений-(DecisionTreeClassifier)-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Модель  Дерево решений (DecisionTreeClassifier)</a></span></li><li><span><a href="#Модель-Случайный-лес-(RandomForestClassifier)" data-toc-modified-id="Модель-Случайный-лес-(RandomForestClassifier)-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Модель Случайный лес (RandomForestClassifier)</a></span></li><li><span><a href="#Модель-Логистическая-регрессия-(LogisticRegression)" data-toc-modified-id="Модель-Логистическая-регрессия-(LogisticRegression)-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Модель Логистическая регрессия (LogisticRegression)</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Проверим-модель-на-тестовой-выборке" data-toc-modified-id="Проверим-модель-на-тестовой-выборке-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Проверим модель на тестовой выборке</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Проверьте-модели-на-адекватность" data-toc-modified-id="Проверьте-модели-на-адекватность-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Проверьте модели на адекватность</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Описание проекта

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

# Описание данных

Каждый объект в наборе данных — это информация о поведении одного пользователя за месяц.  
Известно:  
* сalls — количество звонков,  
* minutes — суммарная длительность звонков в минутах,  
* messages — количество sms-сообщений,  
* mb_used — израсходованный интернет-трафик в Мб,  
* is_ultra — каким тарифом пользовался в течение месяца («Ультра» — 1, «Смарт» — 0).

# Откроем и изучим файл

Импортируем библиотеки

In [1]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.dummy import DummyClassifier

Загрузим данные и посмотри на 10 строк из них

In [2]:
df = pd.read_csv('users_behavior.csv')
df.sample(10)

calls  minutes  messages   mb_used  is_ultra
265    60.0   442.02      16.0  18306.45         1
1775  115.0   890.50      20.0   8643.85         1
1593   23.0   128.42      27.0  16320.86         0
656    45.0   333.64      39.0  16542.45         0
959    48.0   341.80      53.0   5910.35         1
1316   88.0   564.56       0.0  13344.02         0
1786   89.0   594.81      58.0  25825.08         0
2406   53.0   359.13       0.0  13758.39         0
526    26.0   194.24      31.0  11663.69         0
381    82.0   521.66      30.0  13396.91         1

Выведем информацию о таблице

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


Выведем статистическую информацию о таблице

In [4]:
df.describe().T

count          mean          std  min         25%        50%  \
calls     3214.0     63.038892    33.236368  0.0     40.0000     62.000   
minutes   3214.0    438.208787   234.569872  0.0    274.5750    430.600   
messages  3214.0     38.281269    36.148326  0.0      9.0000     30.000   
mb_used   3214.0  17207.673836  7570.968246  0.0  12491.9025  16943.235   
is_ultra  3214.0      0.306472     0.461100  0.0      0.0000      0.000   

                 75%       max  
calls        82.0000    244.00  
minutes     571.9275   1632.06  
messages     57.0000    224.00  
mb_used   21424.7000  49745.73  
is_ultra      1.0000      1.00

Посмотрим на количество и процент пропущенны значений

In [5]:
pd.DataFrame({
    'count_missing': df.isna().sum(),
    'percent_missing': df.isna().mean() * 100
}).sort_values(by='percent_missing', ascending=False)

count_missing  percent_missing
calls                 0              0.0
minutes               0              0.0
messages              0              0.0
mb_used               0              0.0
is_ultra              0              0.0

Посмотрим на количество дубликатов

In [6]:
print('Количество дубликатов {} шт'.format(df.duplicated().sum()))

Количество дубликатов 0 шт


Посмотрим на количество пользователей пользующихся тарифом **ultra**

In [7]:
df['is_ultra'].value_counts()

0    2229
1     985
Name: is_ultra, dtype: int64

## Вывод

После изучения данных о поведении пользователей за месяц, можно сделать вывод, что все столбцы приведены к нужному типу данных, в данных нет пропусков и дубликатов.  
Пользователей с тарифом **smart** больше более чем в 2 раза чем пользователей с тарифом тарифом **ultra**. Из-за этого можем получить разное количество пользователей с тарифом **ultra** при делении данных на разные выборки, необходимо будет это проверить.

# Разобьем данные на выборки

Так как столбец `is_ultra` является категоральным, то перед нами стоит задача **классификации.**   
Разделим данные по признакам и целевой переменной. В данном случае целевой переменной будет столбец `is_ultra`, остальные столбцы будут признаками.

In [8]:
features = df.drop('is_ultra', axis=1)
target = df['is_ultra']

Теперь разобьем данные на обучающую, тестовую и валидационную выборки в соотношении 3:1:1.

In [9]:
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.2, random_state=12345)

features_train, features_valid, target_train, target_valid = train_test_split(
    features_train, target_train, test_size=0.25, random_state=12345)

In [10]:
print('Размер обучающей выборки равен {}, это {:.1%} от данных'.format(
    len(features_train), len(features_train)/features.shape[0]))

print('Размер тестовой выборки равен {}, это {:.1%} от данных'.format(
    len(features_test), len(features_test)/features.shape[0]))

print('Размер валидационной выборки равен {}, это {:.1%} от данных'.format(
    len(features_valid), len(features_valid)/features.shape[0]))

Размер обучающей выборки равен 1928, это 60.0% от данных
Размер тестовой выборки равен 643, это 20.0% от данных
Размер валидационной выборки равен 643, это 20.0% от данных


In [11]:
print(
    'Количество пользователей с тарифом Ultra в обучающей выборке {}шт, это {:.1%} от размера выборки'
    .format(target_train.sum(), target_train.mean()))
print(
    'Количество пользователей с тарифом Ultra в тестовой выборке {}шт, это {:.1%} от размера выборки'
    .format(target_test.sum(), target_test.mean()))
print(
    'Количество пользователей с тарифом Ultra в валидационной выборке {}шт, это {:.1%} от размера выборки'
    .format(target_valid.sum(), target_valid.mean()))

Количество пользователей с тарифом Ultra в обучающей выборке 589шт, это 30.5% от размера выборки
Количество пользователей с тарифом Ultra в тестовой выборке 196шт, это 30.5% от размера выборки
Количество пользователей с тарифом Ultra в валидационной выборке 200шт, это 31.1% от размера выборки


## Вывод

Мы разбили данные на обучающую, тестовую и валидационную выборки в соотношении 3:1:1.  
Пользователей с тарифом Ultra в каждой выборке около 30%.

# Исследуйте модели

Построим несколько моделей:
* Дерево решений (DecisionTreeClassifier)  
* Случайный лес (RandomForestClassifier)  
* Логистическая регрессия (LogisticRegression)

## Модель  Дерево решений (DecisionTreeClassifier)

Создадим цикл с перебором гиперпараметра **max_depth** - максимальная глубина дерева, со значением от 1 до 50 и сохраним модель с лучшим качеством предсказания.

In [12]:
best_tree = None
best_tree_result = 0
best_tree_depth = 0

for depth in range(1, 51):
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model.fit(features_train, target_train)
    prediction = model.predict(features_valid)
    result = accuracy_score(target_valid, prediction)
    if result > best_tree_result:
        best_tree = model
        best_tree_result = result
        best_tree_depth = depth
        
print('Accuracy наилучшей модели дерева решений на валидационной выборке:', round(best_tree_result,3))
print('Максимальная глубина:', best_tree_depth)

Accuracy наилучшей модели дерева решений на валидационной выборке: 0.774
Максимальная глубина: 7


## Модель Случайный лес (RandomForestClassifier)

Создадим двойной цикл с перебором гиперпараметров **n_estimators** - число деревьев в "лесу", со значением от 10 до 50, с шагом 10 и с перебором **max_depth** - максимальная глубина дерева, со значением от 1 до 50 и сохраним модель с лучшим качеством предсказания.

In [14]:
best_forest = None
best_forest_result = 0
best_forest_depth = 0
best_forest_estimators = 0
best_forest_features = None

for estimators in range(10,101,10):
    for depth in range(1, 51):
        for features in ['log2', 'sqrt']:
            model = RandomForestClassifier(n_estimators=estimators, 
                                           max_depth=depth,
                                           max_features = features,
                                           random_state=12345)
            model.fit(features_train, target_train)
            prediction = model.predict(features_valid)
            result = accuracy_score(target_valid, prediction)
            if result > best_forest_result:
                best_forest = model
                best_forest_result = result
                best_forest_depth = depth
                best_forest_estimators = estimators
                best_forest_features = features
            
print('Accuracy наилучшей модели случайного леса на валидационной выборке:',round(best_forest_result,3)) 
print('Максимальная глубина:', best_forest_depth)
print('Количество деревьев:', best_forest_estimators)

Accuracy наилучшей модели случайного леса на валидационной выборке: 0.802
Максимальная глубина: 15
Количество деревьев: 20


## Модель Логистическая регрессия (LogisticRegression)

Создадим модель логистической регрессии, обучим ее и посмотрим на качество ее предсказаний на валидационной выборке

In [15]:
lr_model = LogisticRegression(random_state=12345)
lr_model.fit(features_train, target_train)
lr_prediction = lr_model.predict(features_valid)
lr_result = accuracy_score(target_valid, lr_prediction)
print("Accuracy модели логистической регрессии на валидационной выборке:", round(lr_result,3))

Accuracy модели логистической регрессии на валидационной выборке: 0.726


## Вывод

Мы построили и обучили 3 модели:  
    1) Модель  Дерево решений (DecisionTreeClassifier), показала лучший Accuracy на валидационной выборке равный 0.774 и максимальной глубиной равной 7.  
    2) Модель Случайного леса (RandomForestClassifier), показала лучший Accuracy на валидационной выборке равный 0.802, максимальной глубиной равной 15 и количеством деревьев равным 20.  
    3) Модель Логистической регрессии (LogisticRegression), показала лучший Accuracy на валидационной выборке равный 0.726.
    
Модель случайного леса показала лучший Accuracy среди всех моделей, будем далее использовать её.  
Наихудший результат показала модель Логистической регрессии.

# Проверим модель на тестовой выборке

Посмотрим на предсказания наилучшей моделей случайного леса на тестовой выборке

In [16]:
predict_forest_on_test = best_forest.predict(features_test)
forest_acc_on_test = accuracy_score(target_test, predict_forest_on_test)
print('Accuracy наилучшей модели случайного леса на тестовой выборке:',
      round(forest_acc_on_test, 3))

Accuracy наилучшей модели случайного леса на тестовой выборке: 0.785


## Вывод

Наилучшая модель Случайного леса (RandomForestClassifier) на тестовой выборке получила Accuracy 0.785, почти равный полученному на валидационной. Можно сделать вывод, что нам удалось построить модель для задачи классификации, которая выберет подходящий тариф для пользователей.

# Проверьте модели на адекватность

Проверим модель случайного леса на адекватность с помощью DummyClassifier с гиперпараметром **strategy='most_frequent'**: всегда предсказывает наиболее частую метку в обучающем наборе.

In [17]:
dummy_clf = DummyClassifier(strategy='most_frequent')
dummy_clf.fit(features_train, target_train)
prediction_dummy = dummy_clf.predict(features_test)
dummy_acc = accuracy_score(target_test, prediction_dummy)

print('Accuracy модели случайного леса равна {}, а DummyClassifier равна {}'.
        format(round(forest_acc_on_test,3), round(dummy_acc,3)))

if dummy_acc < forest_acc_on_test:
    print('Модель случайного леса адекватна')
else:
    print('Модель случайного леса не адекватна')

Accuracy модели случайного леса равна 0.785, а DummyClassifier равна 0.695
Модель случайного леса адекватна


## Вывод

В результати проверки модели на адекватность мы получили Accuracy модели случайного леса равной 0.785, а DummyClassifier равной 0.695, что говорит нам о том, что модель случайного леса является адекватной.

# Общий вывод

После изучения данных о поведении пользователей за месяц, можно сделать вывод, что все столбцы приведены к нужному типу данных, в данных нет пропусков и дубликатов.
Пользователей с тарифом smart больше более чем в 2 раза чем пользователей с тарифом тарифом ultra.  
Так как столбец `is_ultra` является категоральным, то перед нами стоит задача классификации.  
Мы разделили данные по признакам и целевой переменной. К целевой переменной мы отнесли столбец `is_ultra`, остальные столбцы отнесли к признаками.  
Далее мы разбили данные на обучающую, тестовую и валидационную выборки в соотношении 3:1:1. И получили пользователей с тарифом Ultra в каждой выборке около 30%.  
После этого мы построили несколько моделей:
* Дерево решений (DecisionTreeClassifier)  
* Случайный лес (RandomForestClassifier)  
* Логистическая регрессия (LogisticRegression)  

Получили слудующие результаты:  
    1) Модель  Дерево решений (DecisionTreeClassifier), показала лучший Accuracy на валидационной выборке равный 0.774 и максимальной глубиной равной 7.    
    2) Модель Случайного леса (RandomForestClassifier), показала лучший Accuracy на валидационной выборке равный 0.802, максимальной глубиной равной 15 и количеством деревьев равным 20.    
    3) Модель Логистической регрессии (LogisticRegression), показала лучший Accuracy на валидационной выборке равный 0.726.  
    
Модель случайного леса показала лучший Accuracy среди всех моделей, поэтому было принято решение далее использовать её. Наихудший результат показала модель Логистической регрессии.

Далее мы проверили модель случайного леса на тестовой выборке. Наилучшая модель Случайного леса (RandomForestClassifier) на тестовой выборке получила Accuracy 0.785, почти равный полученному на валидационной. Можно сделать вывод, что нам удалось построить модель для задачи классификации, которая выберет подходящий тариф для пользователей.  

Для того, что бы проверить модель случайного леса на адекватность мы использовали DummyClassifier с гиперпараметром **strategy='most_frequent'**: всегда предсказывает наиболее частую метку в обучающем наборе. В результати проверки модели на адекватность мы получили Accuracy модели случайного леса равной 0.785, а DummyClassifier равной 0.695, что говорит нам о том, что модель случайного леса является адекватной.